gnomad_constraint/utils/constraint.py

```diff
@@ -81,3 +88,101 @@ def prepare_ht_for_constraint_calculations(ht: hl.Table) -> hl.Table:
+
+    The returned Table includes the following annotations:
+        - context - trinucleotide genomic context
+        - ref - the reference allele
+        - alt - the alternate base
+        - methylation_level - methylation_level
+        - observed_variants - observed variant counts in `exome_ht`
+        - possible_variants - possible variant counts in `context_ht`
+        - downsampling_counts_{pop} - variant counts in downsamplings for populations in `pops`
+        - mu_snp - SNP mutation rate
+        - annotations added by `annotate_mutation_type`
+
+    :param exome_ht: Preprocessed exome Table.
+    :param context_ht: Preprocessed context Table.
+    :param mutation_ht: Preprocessed mutation rate Table.
+    :param max_af: Maximum allele frequency for a variant to be included in returned counts. Default is 0.001.
+    :param keep_annotations: Annotations to keep in the context Table.
+    :param pops: List of populations to use for downsampling counts. Default is ().
+    :param grouping: Annotations other than 'context', 'ref', 'alt', and `methylation_level` to group by when counting variants. Default is ('exome_coverage',).
+    :param partition_hint: Target number of partitions for aggregation. Default is 100.
+    :return: Table with observed variant and possible variant count.
+    """
+    # Allele frequency information for high-quality genotypes (GQ >= 20; DP >= 10; and AB >= 0.2 for heterozygous calls) in all release samples in gnomAD.
+    freq_expr = exome_ht.freq[0]
+
+    # Set up the criteria to exclude variants not observed in the dataset, low-quality variants, and variants with allele frequency
+    # above the `max_af` cutoff.
+    keep_criteria = (
+        (freq_expr.AC > 0) & exome_ht.pass_filters & (freq_expr.AF <= max_af)
+    )
+    keep_annotations += grouping
+
+    # Keep variants that are synonymous, canonical, and satisfy the criteria above.
+    # Count the observed variants in the entire Table and in each downsampling grouped by `keep_annotations`.
+    ht = count_variants_by_group(
+        filter_vep_transcript_csqs(exome_ht.filter(keep_criteria)).select(
+            *list(keep_annotations) + ["freq", "pass_filters"]
+        ),
+        additional_grouping=grouping,
+        partition_hint=partition_hint,
+        count_downsamplings=pops,
+        use_table_group_by=True,
+    )
+    ht = ht.transmute(observed_variants=ht.variant_count)
+
+    # Keep variants that are synonymous and canonical in `context_ht`.
+    context_ht = filter_vep_transcript_csqs(context_ht).select(*keep_annotations)
+    # Filter the `exome_ht` to rows that don’t match the criteria above
+    # Anti join the `context_ht` with filtered `exome_ht`, so that `context_ht` only has rows that match the
+    # criteria above in the `exome_ht` or are never in the `exome_ht`.
```

Comment:
```suggestion
    # Anti join the `context_ht` with filtered `exome_ht`, so that `context_ht` only
    # has rows that match the criteria above in the `exome_ht` or are never in
    # the `exome_ht`.
```